File: Javascript/2. Add Two Numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let tail;
    let head;
    let carry = 0;
    
    while (l1 || l2 || carry) {
        let sum = carry;
        if(l1) {
            sum += l1.val;
            l1 = l1.next;
        }
        if(l2) {
            sum += l2.val;
            l2 = l2.next;
        }
        
        carry = Math.floor(sum / 10);
        let node = new ListNode(sum % 10);
        
        if (!head) {
            tail = node;
            head = tail;
        } else {
            tail.next = node
            tail = tail.next
        }
    }

    return head
};

File: 2. Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *res = new ListNode(0);
        ListNode *p = res;
        int adv = 0;
        
        while(l1 != NULL || l2 != NULL || adv != 0) {
            int ans = adv;
            if(l1 != NULL) {
                ans += l1->val;
                l1 = l1->next;
            }
            if(l2 != NULL) {
                ans += l2->val;
                l2 = l2->next;
            }
            
            p->next = new ListNode(ans % 10);
            p = p->next;
            adv = ans / 10;
        }
        
        return res->next;
    }
};
